### ISSUE TITLE
Subscriber API is (technically) unsound

### ISSUE THREAD

**Issue by Ekleog** [tokio-rs/tracing#1171]
### Version & Crates

On both tags tracing-0.1.22 / tracing-core-0.1.17 on the git repository, in file `tracing-core/src/subscriber.rs`

### Description

Function `downcast_raw` of trait `tracing_core::Subscriber` is `unsafe` in a safe trait.

However, `unsafe` code relies on the implementation of this function to be correct, as mentioned by the `Safety` blurb in the documentation comments.

Rust rules state that a `unsafe fn` depict a requirement that must be handled by the caller, and an `unsafe` block a guarantee that the requirements are verified. This is not what is wanted here. Instead, what would accurately describe the requirements here would be `unsafe trait`, that depicts a requirement that must be handled by the trait implementer, and then `unsafe impl` would be the guarantee of the implementer that the guarantees are verified.

Now… this will be a lot of churn, and in practice it probably doesn't hurt much to have only `downcast_raw` be `unsafe`. So the ways forward I see are:
- Ignore the problem, as in practice it's likely that people will understand what this means (though technically the trait is unsound)
- Ignore the problem for now, make `Subscriber` `unsafe` for the next semver-breaking version
- Stop actually using `downcast_raw` from `unsafe` code that relies on its implementation being correct, in which case the function becomes useless to implement but the `Subscriber` API is no longer unsound

PS: I must say that when I say that this `Subscriber` API is technically unsound, it is to mean that no meaningful use of `downcast_ref` can happen: technically the UB is in the implementation of `downcast_ref` for `dyn Subscriber`, but this only stems from the API of `Subscriber` not being able to hold up its guarantees due to the `unsafe` being placed at the wrong place.

Anyway, this is probably not a big deal, but I felt like I had to report it.
Thank you for doing this crate, it's pretty much great!

**Comment by hawkw** [comment]
I don't think it makes sense for the `Subscriber` trait itself to be unsafe to implement; since `downcast_raw` has a default implementation, a majority of `Subscriber` implementations don't override its behavior. You're right, though, that technically, it would be correct to make the trait unsafe.

I'm not sure if there's any solution for this that makes sense besides strengthening the notes in the documentation that this method is unsafe to implement rather than to call. Ideally, the language would give us some way of differentiating between this without making the rest of the trait unsafe to implement.

**Comment by MikailBag** [comment]
As a workaround:
```rust
pub struct Wrapper(Option<*const ()>);

impl Wrapper {
    pub unsafe fn new() -> Wrapper;

    pub fn into_inner(self) -> *const ();
}

trait Subscriber {
   // ...
   fn downcast_raw(&self, ty: TypeId) -> Wrapper;
   // ...
}
```

**Comment by hawkw** [comment]
> As a workaround:
> 
> ```rust
> pub struct Wrapper(Option<*const ()>);
> 
> impl Wrapper {
>     pub unsafe fn new() -> Wrapper;
> 
>     pub fn into_inner(self) -> *const ();
> }
> 
> trait Subscriber {
>    // ...
>    fn downcast_raw(&self, ty: TypeId) -> Wrapper;
>    // ...
> }
> ```

Yeah, I'm considering something like this in `tracing` 0.2...it also has the nice side benefit that the `into_inner` method actually _doesn't_ need to be `pub`, so you are also essentially preventing user code for using the returned raw pointer to construct a reference that might outlive the `self` ref, which is also good. Of course, it's a breaking change, so it'll have to wait.
